### PR TITLE
feat - use fixture plugins instead of conftest files

### LINKED ISSUE #2095: `tests/conftest.py` is getting too crowded
## Description

`tests/conftest.py` has too many fixtures that **need** to be there because they are accessed by tests in different locations. The file has some comments with hashtags that try to divide it into sections but is not getting the readability easier.

## Suggested solution

SD: Group the tests by functionality or application (e.g. demultiplex, timestamps, HK) into scripts accessible by `tests/conftest.py` in the following way:
```shell
tests/
├── conftest.py
├── fixtures/
│   ├── __init__.py
│   ├── hk.py
│   └── demultiplex.py
```

The file `tests/conftest.py` should look like this:
```python
pytest_plugins = [
    "fixtures.hk",
    "fixtures.demultiplex",
]
```

## This can be closed when

The file `tests/conftest.py` has been divided into different files in `tests/fixtures/<fixture_topic>.py`.

## Blocked by

*If there are any blocking issues/prs/things in this or other repos. Please link to them.*

### PR DESCRIPTION
**Pull request by diitaz93**
## Description
Fix #2095 and continue #2259. 

### TL;DR
The main `conftest.py` was too crowded, now the fixtures are separated into different modules available to all tests.

### Details 
The current test system looks for fixtures defined in the `conftest.py` files, either in the local directory or in any directory above, but it is impossible to use a fixture defined in a directory below it. For example, a test function in `test_module_top.py` can't access a fixture defined in `tests/subfolder/subsubfolder/conftest.py`:
```shell
tests
├── conftest.py
├── subfolder
│   ├── sub_subfolder
│   │   ├── __init__.py
│   │   ├── conftest.py
│   │   └── test_module_sub.py
└── test_module_top.py
```
To solve this, we started writing all our fixtures in the first contest, i.e. `tests/conftest.py`. This made this file extremely crowded and difficult to navigate.

This PR changes the conftest structure for a plugin structure, in which the fixtures are grouped by topic in a folder called `fixture_plugins` and imported in the conftest as a plugin. In this way, the structure of the tests is
```shell
tests
├── conftest.py
├── fixture_plugins
│   ├── fixture_subfolder
│   │   ├── fixtures_topic_1.py
│   │   └── __init__.py
│   ├── fixtures_topic_2.py
│   └── __init__.py
├── test_subfolder
│   ├── __init__.py
│   └── test_module_sub.py
└── test_module_top.py
```
and both `test_module_top.py` and `test_module_sub.py` have access to all the fixtures defined in `fixture_plugins`.

### Added

- Fixture plugin modules with fixtures in the new folder `tests/fixture_plugins/` with the following structure:
```shell
tests
├── conftest.py
├── fixture_plugins
│   ├── demultiplex_fixtures
│   │   ├── __init__.py
│   │   ├── flow_cell_fixtures.py
│   │   ├── name_fixtures.py
│   │   ├── path_fixtures.py
│   │   ├── run_parameters_fixtures.py
│   │   └── sample_fixtures.py
│   ├── __init__.py
│   └── timestamp_fixtures.py
```

### Changed

- Moved demultiplexing and timestamp fixtures from `conftest.py` to separate modules in the new fixture plugin directory.

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by seallard** [comment]
Nice initiative! Is this a common pattern? And can we still rely on pytests "auto" import of fixtures?

**Comment by ChrOertlin** [comment]
It looks nicely structured and seems like to improve the ability to lookup fixtures more easily. It would be great to expand this ot other modules. If we can retain all other functionality, maybe the tests/conftest.py will end up containing only imports.

**Comment by islean** [comment]
I love it! As long as the auto imports work, I love it.

**Review by henrikstranneheim — COMMENTED** [review]
Looks like a great improvement!

**Comment by diitaz93** [comment]
> Nice initiative! Is this a common pattern? And can we still rely on pytests "auto" import of fixtures?

I know requesting plugins in the root contest is a common approach, but honestly I am not sure how common it is to request fixture. The documentation does mention fixtures but very briefly https://docs.pytest.org/en/7.2.x/how-to/writing_plugins.html#requiring-loading-plugins-in-a-test-module-or-conftest-file.

We can rely on the fixtures to be loaded automatically, it is equivalent to qritting them down on the root conftest

**Comment by diitaz93** [comment]
> It looks nicely structured and seems like to improve the ability to lookup fixtures more easily. It would be great to expand this ot other modules. If we can retain all other functionality, maybe the tests/conftest.py will end up containing only imports.

Exactly. It is still left to discuss if we would like a hybrid approach with some fixtures in the contests (that could be very specific and only used in one test) or a fixture structure purely based on plugin requests

**Comment by seallard** [comment]
> > It looks nicely structured and seems like to improve the ability to lookup fixtures more easily. It would be great to expand this ot other modules. If we can retain all other functionality, maybe the tests/conftest.py will end up containing only imports.
> 
> Exactly. It is still left to discuss if we would like a hybrid approach with some fixtures in the contests (that could be very specific and only used in one test) or a fixture structure purely based on plugin requests

Just to be clear, this pattern does not force us to import fixtures directly in files? If that is the case I think we should consider other options

**Comment by diitaz93** [comment]
The full migration to plugins will be done in several PRs

File: tests/fixture_plugins/demultiplex_fixtures/flow_cell_fixtures.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+def novaseq_6000_pre_1_5_kits_flow_cell_data(flow_cells_dir: Path) -> FlowCellDirectoryData:
+    return FlowCellDirectoryData(Path(flow_cells_dir, "190927_A00689_0069_BHLYWYDSXX"))
+
+
+@pytest.fixture()
+def novaseq_x_flow_cell_data(flow_cells_dir: Path) -> FlowCellDirectoryData:
+    return FlowCellDirectoryData(Path(flow_cells_dir, "20231108_LH00188_0028_B22F52TLT3"))
+
+
+# Broken flow cells
```
Comment (ChrOertlin) [inline tests/fixture_plugins/demultiplex_fixtures/flow_cell_fixtures.py:59]: maybe the docstring can contain why a flow cell is broken within the docstring or the actual fixute name so we can better distinguish them?

File: tests/fixture_plugins/demultiplex_fixtures/name_fixtures.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+def novaseq_x_flow_cell_full_name() -> str:
+    """Return the full name of a NovaSeqX flow cell."""
+    return "20230508_LH00188_0003_A22522YLT3"
+
+
+# Lists
+
+
+@pytest.fixture(scope="session")
+def bcl_convert_demultiplexed_flow_cell_sample_internal_ids() -> list[str]:
```
Comment (ChrOertlin) [inline tests/fixture_plugins/demultiplex_fixtures/name_fixtures.py:102]: nice!

File: cg/constants/demultiplexing.py [hunk 3]
```diff
@@ -29,6 +29,7 @@ class DemultiplexingDirsAndFiles(StrEnum):
     RUN_PARAMETERS_PASCAL_CASE: str = "RunParameters.xml"
     RUN_PARAMETERS_CAMEL_CASE: str = "runParameters.xml"
     SAMPLE_SHEET_FILE_NAME: str = "SampleSheet.csv"
+    CORRECT_SAMPLE_SHEET: str = "CorrectSampleSheet.csv"
```
Comment (Vince-janv) [inline cg/constants/demultiplexing.py:32]: There is never a file generated called `CorrectSampleSheet.csv`. It is only there for testing purposes. And production code should not be adapted to suit the tests

**Review by ChrOertlin — APPROVED** [review]
I really like the new structure, it is much easier to see and understand what fixtures we are working with!

File: tests/fixture_plugins/demultiplex_fixtures/flow_cell_fixtures.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+from cg.constants.demultiplexing import BclConverter, DemultiplexingDirsAndFiles
+from cg.models.flow_cell.flow_cell import FlowCellDirectoryData
+
+# Functional flow cells
+
+
+@pytest.fixture(scope="module")
+def hiseq_x_single_index_flow_cell(
+    hiseq_x_single_index_flow_cell_dir: Path,
```
Comment (Vince-janv) [inline tests/fixture_plugins/demultiplex_fixtures/flow_cell_fixtures.py:14]: Out of curiosity how is `hiseq_x_single_index_flow_cell_dir` found? Is it because it's in the same directory as this file?

**Review by Vince-janv — APPROVED** [review]
This looks great! ⭐

File: cg/constants/demultiplexing.py (hunk 3 above)
Comment (diitaz93) [reply cg/constants/demultiplexing.py:32]: Good catch, I will move this

File: tests/fixture_plugins/demultiplex_fixtures/flow_cell_fixtures.py (hunk 4 above)
Comment (diitaz93) [reply tests/fixture_plugins/demultiplex_fixtures/flow_cell_fixtures.py:14]: That fixture is defined in `tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py`, which is called as a plugin in the root `conftest.py`. It is equivalent to defining the fixture in the contest, so it is available to all tests.

File: tests/fixture_plugins/demultiplex_fixtures/flow_cell_fixtures.py (hunk 1 above)
Comment (diitaz93) [reply tests/fixture_plugins/demultiplex_fixtures/flow_cell_fixtures.py:59]: yes, that is the purpose of my next PR. My idea is that many of these fixtures can be removed or replaced by existing ones. The fixtures for functional flow cells should be the ones above and no more, because those are the flow cells that are well-documented.
